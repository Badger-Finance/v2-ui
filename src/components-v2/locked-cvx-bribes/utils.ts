import BigNumber from 'bignumber.js';
import { formatWithoutExtraZeros, numberWithCommas, unscale } from '../../mobx/utils/helpers';

export function formatBalance(balance: BigNumber): string {
  return numberWithCommas(formatWithoutExtraZeros(unscale(balance, 18)));
}
