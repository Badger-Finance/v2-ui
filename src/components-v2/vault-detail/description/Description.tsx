import { VaultDTO } from '@badger-dao/sdk';
import { Grid, Typography } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import React from 'react';

import VaultLogo from '../../landing/VaultLogo';

const useStyles = makeStyles((theme) => ({
	root: {
		display: 'flex',
	},
	namesContainer: {
		marginLeft: theme.spacing(1),
	},
	settName: {
		display: 'inline',
		fontSize: 24,
	},
	vaultName: {
		fontSize: 14,
	},
	settLogo: {
		width: '100%',
		margin: 'auto',
	},
	logoContainer: {
		display: 'flex',
		alignItems: 'center',
	},
}));

interface Props {
	vault: VaultDTO;
}

export const Description = ({ vault }: Props): JSX.Element => {
	const classes = useStyles();

	return (
		<div className={classes.root}>
			<Grid item className={classes.logoContainer}>
				<VaultLogo tokens={vault.tokens} />
			</Grid>
			<Grid item container direction="column" justifyContent="center" className={classes.namesContainer}>
				<Grid item container alignItems="center">
					<Typography className={classes.settName}>
						{vault.protocol} - {vault.name}
					</Typography>
				</Grid>
				<Grid item>
					<Typography className={classes.vaultName} color="textSecondary">
						{vault.asset}
					</Typography>
				</Grid>
			</Grid>
		</div>
	);
};
