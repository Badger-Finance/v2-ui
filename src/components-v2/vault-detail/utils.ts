import { Vault, VaultData } from '@badger-dao/sdk';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/vaults/vault-charts';
import { DelaySeverity } from '../../mobx/model/vaults/vault-rewards';

dayjs.extend(utc);

export const ChartModeTitles: Record<string, string> = {
	[ChartMode.Value]: 'Sett Value',
	[ChartMode.Ratio]: 'Token Ratio',
	[ChartMode.AccountBalance]: 'My Holdings',
	[ChartMode.BoostMultiplier]: 'Badger Boost',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};

export function defaultVaultBalance(vault: Vault): VaultData {
	return {
		address: vault.vaultToken,
		name: vault.name,
		symbol: vault.asset,
		pricePerFullShare: vault.pricePerFullShare,
		balance: 0,
		value: 0,
		earnedBalance: 0,
		earnedValue: 0,
		depositedBalance: 0,
		withdrawnBalance: 0,
		tokens: [],
		earnedTokens: [],
	};
}

export interface Balance {
	balance: number;
}

export function hasBalance(input?: Balance): boolean {
	if (!input) {
		return false;
	}
	return input.balance > 0;
}
