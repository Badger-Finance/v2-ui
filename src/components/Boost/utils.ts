import { makeStyles } from '@material-ui/core/styles';
import { ClassNameMap } from '@material-ui/core/styles/withStyles';
import React from 'react';

type ComparisonConfig = {
  toCompareValue: number;
  toBeComparedValue: number;
  greaterCaseColor: string;
  lessCaseColor: string;
  defaultColor: string;
};

export const getColorFromComparison = ({
  toCompareValue,
  toBeComparedValue,
  greaterCaseColor,
  lessCaseColor,
  defaultColor,
}: ComparisonConfig): string => {
  if (toCompareValue > toBeComparedValue) return greaterCaseColor;
  if (toCompareValue < toBeComparedValue) return lessCaseColor;

  return defaultColor;
};

export const useAssetInputStyles = (
  currentValue: string,
  toCompare = 0,
): ((props?: React.CSSProperties) => ClassNameMap<'assetColor'>) => {
  return makeStyles((theme) => {
    const defaultColor = currentValue ? theme.palette.text.primary : theme.palette.text.secondary;
    const fontColor = getColorFromComparison({
      toCompareValue: Number(currentValue),
      toBeComparedValue: toCompare,
      greaterCaseColor: '#74D189',
      lessCaseColor: theme.palette.error.main,
      defaultColor,
    });

    return {
      assetColor: {
        color: fontColor,
      },
    };
  });
};
