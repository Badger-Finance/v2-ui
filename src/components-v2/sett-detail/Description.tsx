import React from 'react';
import { Box, Grid, Typography } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import TrendingUpIcon from '@material-ui/icons/TrendingUp';
import TrendingDownIcon from '@material-ui/icons/TrendingDown';
import { Sett } from '../../mobx/model/setts/sett';
import { Performance } from '../../mobx/model/rewards/performance';
import clsx from 'clsx';
import { formatWithoutExtraZeros } from '../../mobx/utils/helpers';

const reduceSourcePerformance = (prev: Performance, current: Performance) => {
	const {
		oneDay: prevOneDay = 0,
		threeDay: prevThreeDay = 0,
		sevenDay: prevSevenDay = 0,
		thirtyDay: prevThirtyDay = 0,
	} = prev;

	const {
		oneDay: currentOneDay = 0,
		threeDay: currentThreeDay = 0,
		sevenDay: currentSevenDay = 0,
		thirtyDay: currentThirtyDay = 0,
	} = current;

	return {
		oneDay: prevOneDay + currentOneDay,
		threeDay: prevThreeDay + currentThreeDay,
		sevenDay: prevSevenDay + currentSevenDay,
		thirtyDay: prevThirtyDay + currentThirtyDay,
	};
};

const getSourcesPerformanceSummary = (sett: Sett) => {
	return sett.sources.map((source) => source.performance).reduce(reduceSourcePerformance);
};

const useStyles = makeStyles((theme) => ({
	root: {
		display: 'flex',
	},
	namesContainer: {
		marginLeft: theme.spacing(1),
	},
	settName: {
		display: 'inline',
		fontSize: 20,
	},
	vaultName: {
		fontSize: 14,
	},
	settLogo: {
		width: '100%',
		margin: 'auto',
	},
	logoContainer: {
		display: 'flex',
		width: 50,
		height: 50,
	},
	apyBadge: {
		display: 'inline-flex',
		alignItems: 'center',
		maxHeight: 24,
		fontSize: 12,
		borderRadius: 4,
		padding: 4,
		color: theme.palette.common.black,
		marginLeft: theme.spacing(1),
	},
	apyText: {
		marginLeft: 4,
	},
	increasedApy: {
		backgroundColor: '#74D189',
	},
	reducedApy: {
		backgroundColor: theme.palette.error.main,
	},
}));

interface Props {
	sett: Sett;
}

export const Description = ({ sett }: Props): JSX.Element => {
	const classes = useStyles();
	const nameHasSpaces = sett.name.split(' ').length > 1;
	const displayName = nameHasSpaces ? sett.name.split(' ').slice(1).join(' ') : sett.name;

	const { sevenDay = 0 } = getSourcesPerformanceSummary(sett);

	// TODO decide if we want to include multiple change criteria, i.e: difference las 24rs, difference las week
	const sevenDaysPerformanceComparison = sett.apr - sevenDay;
	const isPositiveDifference = sevenDaysPerformanceComparison > 0;
	const shouldShowDifferenceBadge = sevenDaysPerformanceComparison !== 0; // only show when there's actual change

	return (
		<div className={classes.root}>
			<Grid item className={classes.logoContainer}>
				<img
					className={classes.settLogo}
					src={`/assets/icons/${sett.asset.toLowerCase()}.png`}
					alt={`Badger ${sett.name} Vault Symbol`}
				/>
			</Grid>
			<Grid item className={classes.namesContainer}>
				<Box display="flex" alignItems="center">
					<Typography className={classes.settName}>{displayName}</Typography>
					{shouldShowDifferenceBadge && (
						<Typography
							className={clsx(
								classes.apyBadge,
								isPositiveDifference ? classes.increasedApy : classes.reducedApy,
							)}
						>
							{isPositiveDifference ? <TrendingUpIcon /> : <TrendingDownIcon />}
							<span className={classes.apyText}>
								{`${formatWithoutExtraZeros(Math.abs(sevenDaysPerformanceComparison), 2)}%`}
							</span>
						</Typography>
					)}
				</Box>
				<Typography className={classes.vaultName} color="textSecondary">
					{sett.asset}
				</Typography>
			</Grid>
		</div>
	);
};
