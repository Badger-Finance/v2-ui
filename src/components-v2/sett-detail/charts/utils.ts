import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { Sett } from '../../../mobx/model/setts/sett';
import { fetchSettChartInformation } from '../../../mobx/utils/apiV2';
import { SettSnapshotGranularity } from '../../../mobx/model/setts/sett-snapshot';
import { SettChartData } from '../../../mobx/model/setts/sett-charts';

dayjs.extend(utc);

export enum SettChartTimeframe {
	'day' = 1,
	'week' = 7,
	'month' = 30,
}

/**
 * Fetches chart information since the provided timeframe until the current date
 * @param sett
 * @param timeframe
 */
export const fetchSettChart = async (sett: Sett, timeframe: SettChartTimeframe): Promise<SettChartData[] | null> => {
	const timeframeDays = 1 * timeframe;
	const isDayTimeFrame = timeframe === SettChartTimeframe.day;

	// if timeframe is just one day then we want the granularity to be hours
	const granularity = isDayTimeFrame ? SettSnapshotGranularity.HOUR : SettSnapshotGranularity.DAY;

	const now = dayjs().utc(); // query until current date
	const from = dayjs(now).subtract(timeframeDays, 'days').utc();

	const fetchedData = await fetchSettChartInformation(sett.vaultToken, from.toDate(), now.toDate(), granularity);

	if (!fetchedData) {
		return null;
	}

	// data needs to be ascending sorted
	return fetchedData
		.sort((a, b) => a.timestamp - b.timestamp)
		.map((d) => ({ ...d, timestamp: new Date(d.timestamp) }));
};
