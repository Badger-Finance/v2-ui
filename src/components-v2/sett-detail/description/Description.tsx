import React from 'react';
import { Box, Grid, Typography } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { Performance, scalePerformance } from '../../../mobx/model/rewards/performance';
import ApyDisplayBadge, { ComparisonMode } from './ApyComparisonBadge';
import { formatWithoutExtraZeros } from '../../../mobx/utils/helpers';
import { ApyComparisonModeSelector } from './ApyComparisonModeSelector';
import { Vault } from '@badger-dao/sdk';

const reduceSourcePerformance = (prev: Performance, current: Performance) => {
  const {
    oneDay: prevOneDay = 0,
    threeDay: prevThreeDay = 0,
    sevenDay: prevSevenDay = 0,
    thirtyDay: prevThirtyDay = 0,
  } = prev;

  const {
    oneDay: currentOneDay = 0,
    threeDay: currentThreeDay = 0,
    sevenDay: currentSevenDay = 0,
    thirtyDay: currentThirtyDay = 0,
  } = current;

  return {
    oneDay: prevOneDay + currentOneDay,
    threeDay: prevThreeDay + currentThreeDay,
    sevenDay: prevSevenDay + currentSevenDay,
    thirtyDay: prevThirtyDay + currentThirtyDay,
  };
};

const getSourcesPerformanceSummary = (sett: Vault) => {
  const initialValue = { oneDay: 0, threeDay: 0, sevenDay: 0, thirtyDay: 0 };
  return sett.sources
    .map((source) => {
      // reduce sources to baseline - zero boost definitions
      const { boostable, minApr, apr, performance } = source;
      if (boostable) {
        const performanceScalar = minApr / apr;
        source.performance = scalePerformance(performance, performanceScalar);
      }
      return source;
    })
    .map((source) => source.performance)
    .reduce(reduceSourcePerformance, initialValue);
};

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
  },
  namesContainer: {
    marginLeft: theme.spacing(1),
  },
  settName: {
    display: 'inline',
    fontSize: 24,
  },
  vaultName: {
    fontSize: 14,
  },
  settLogo: {
    width: '100%',
    margin: 'auto',
  },
  logoContainer: {
    display: 'flex',
    width: 68,
    height: 68,
  },
}));

interface Props {
  sett: Vault;
}

export const Description = ({ sett }: Props): JSX.Element => {
  const [timeframe, setTimeframe] = React.useState<keyof Performance>('sevenDay');
  const classes = useStyles();
  const performanceSummary = getSourcesPerformanceSummary(sett);
  const performance = performanceSummary[timeframe] ?? 0;

  let performanceResultMode = ComparisonMode.neutral;
  if (performance > 0) {
    performanceResultMode = ComparisonMode.positive;
  } else if (performance < 0) {
    performanceResultMode = ComparisonMode.negative;
  }

  return (
    <div className={classes.root}>
      <Grid item className={classes.logoContainer}>
        <img
          className={classes.settLogo}
          src={`/assets/icons/${sett.vaultAsset.toLowerCase()}.png`}
          alt={`Badger ${sett.name} Vault Symbol`}
        />
      </Grid>
      <Grid item container direction="column" justifyContent="center" className={classes.namesContainer}>
        <Grid item container alignItems="center">
          <Grid item>
            <Typography className={classes.settName}>{sett.name}</Typography>
          </Grid>
          <Grid item>
            <Box display="flex" alignItems="center">
              <ApyDisplayBadge
                apyComparison={`${formatWithoutExtraZeros(Math.abs(performance), 2)}%`}
                mode={performanceResultMode}
              />
            </Box>
          </Grid>
          <Grid item>
            <ApyComparisonModeSelector value={timeframe} onChange={setTimeframe} />
          </Grid>
        </Grid>
        <Grid item>
          <Typography className={classes.vaultName} color="textSecondary">
            {sett.asset}
          </Typography>
        </Grid>
      </Grid>
    </div>
  );
};
