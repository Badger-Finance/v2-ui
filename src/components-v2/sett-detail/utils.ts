import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/setts/sett-charts';
import { DelaySeverity } from '../../mobx/model/setts/sett-rewards';
import { SettDetailMode } from '../../mobx/model/setts/sett-detail';

dayjs.extend(utc);

export const ChartModeTitles = {
	[ChartMode.value]: 'Sett Value',
	[ChartMode.ratio]: 'Token Ratio',
	[ChartMode.accountBalance]: 'Your Holdings',
};

export const SettModeTitles = {
	[SettDetailMode.settInformation]: 'Vault Info',
	[SettDetailMode.userInformation]: 'User Info',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};
