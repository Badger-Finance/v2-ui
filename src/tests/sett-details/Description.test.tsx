import React from 'react';
import { checkSnapshot } from 'tests/utils/snapshots';

import { Description } from '../../components-v2/vault-detail/description/Description';
import { SAMPLE_VAULT } from '../utils/samples';

describe('Description', () => {
  it('displays sett description', () => {
    checkSnapshot(<Description vault={SAMPLE_VAULT} />);
  });
});
