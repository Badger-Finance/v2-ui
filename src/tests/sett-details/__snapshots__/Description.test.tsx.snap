// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Description displays sett description 1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <div
      class="MuiGrid-root-9 makeStyles-logoContainer-6 MuiGrid-item-11"
    >
      <img
        class="makeStyles-settLogo-5"
        src="/assets/icons/slp-digg-wbtc.png"
      />
    </div>
    <div
      class="MuiGrid-root-9 makeStyles-namesContainer-2 MuiGrid-item-11"
    >
      <div
        class="MuiBox-root-112 MuiBox-root-113"
      >
        <p
          class="MuiTypography-root-114 makeStyles-settName-3 MuiTypography-body1-116"
        >
          Wrapped BTC/DIGG
        </p>
        <p
          class="MuiTypography-root-114 makeStyles-apyBadge-7 MuiTypography-body1-116"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-144"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16 6l2.29 2.29-4.88 4.88-4-4L2 16.59 3.41 18l6-6 4 4 6.3-6.29L22 12V6z"
            />
          </svg>
          <span
            class="makeStyles-apyText-8"
          >
            139.34%
          </span>
        </p>
      </div>
      <p
        class="MuiTypography-root-114 makeStyles-vaultName-4 MuiTypography-body1-116 MuiTypography-colorTextSecondary-140"
      >
        SLP-DIGG-WBTC
      </p>
    </div>
  </div>
</div>
`;
